markovflow/models/variational_cvi_sde.py

```diff
@@ -0,0 +1,348 @@
+                tf.zeros((self.output_dim, self.state_dim - 1), dtype=default_float()),
+            ],
+            axis=-1,
+        )
+        # tile for each time point
+        # expand emission_matrix from [output_dim, state_dim], to [1, 1 ... output_dim, state_dim]
+        # where there is a singleton dimension for the dimensions of time points
+        batch_shape = time_points.shape[:-1]  # num_data may be undefined so skip last dim
+        shape = tf.concat(
+            [tf.ones(len(batch_shape) + 1, dtype=tf.int32), tf.shape(emission_matrix)], axis=0
+        )
+        emission_matrix = tf.reshape(emission_matrix, shape)
+
+        # tile the emission matrix into shape batch_shape + [num_data, output_dim, state_dim]
+        repetitions = tf.concat([tf.shape(time_points), [1, 1]], axis=0)
+        return EmissionModel(tf.tile(emission_matrix, repetitions))
+
+    @property
+    def posterior_kalman(self) -> KalmanFilterWithSparseSites:
+        """Build the Kalman filter object from the prior state space models and the sites."""
+        return KalmanFilterWithSparseSites(
+            state_space_model=self.dist_p,
+            emission_model=self.generate_emission_model(tf.reshape(self.time_points, (-1))),
+            sites=self.sites,
+            num_grid_points=self.grid.shape[0],
+            observations_index=self.obs_sites_indices,
+            observations=self._observations
+        )
+
+    @property
+    def posterior(self):
+        raise NotImplementedError
+
+    def local_objective(self, Fmu: TensorType, Fvar: TensorType, Y: TensorType) -> TensorType:
+        """
+        Calculate local loss in CVI.
+
+        :param Fmu: Means with shape ``[..., latent_dim]``.
+        :param Fvar: Variances with shape ``[..., latent_dim]``.
+        :param Y: Observations with shape ``[..., observation_dim]``.
+        :return: A local objective with shape ``[...]``.
+        """
+        return self._likelihood.variational_expectations(Fmu, Fvar, Y)
+
+    def local_objective_and_gradients(self, Fmu: TensorType, Fvar: TensorType) -> [TensorType, TensorType]:
+        """
+        Return the local objective and its gradients with regard to the expectation parameters.
+
+        :param Fmu: Means :math:`μ` with shape ``[..., latent_dim]``.
+        :param Fvar: Variances :math:`σ²` with shape ``[..., latent_dim]``.
+        :return: A local objective and gradient with regard to :math:`[μ, σ² + μ²]`.
+        """
+        with tf.GradientTape() as g:
+            g.watch([Fmu, Fvar])
+            local_obj = tf.reduce_sum(
+                input_tensor=self.local_objective(Fmu, Fvar, self._observations)
+            )
+        grads = g.gradient(local_obj, [Fmu, Fvar])
+        # turn into gradient wrt μ, σ² + μ²
+        grads = gradient_transformation_mean_var_to_expectation((Fmu, Fvar), grads)
+
+        return local_obj, grads
+
+    def cross_term(self, dist_q: StateSpaceModel = None) -> TensorType:
```

Comment:
maybe it would be good to call this kl_cross_term
and to remind the kl expression in the docstrings and how it split.
I m thinking about the other people who might read this in the future

Reply:
Makes sense! I will rename the function and modify the docstring